### PR TITLE
Hclang 116 type rules

### PR DESCRIPTION
**Pull request by Lambcreek**
### Description
Added explanation of HCL type rules to the report

### Changelog
 - Added Type Rules section to report
 - Added command to most .tex file so TeXStudio allows you to compile report without having main file open.

### Issues solved
 - HCLANG-116

### Issues opened as a result of this 
Add something here if new issues arise because of this. For instance if new features-requests or bugs arise from this PR. If this is not the case, then remove this section.

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/3.Theory/TypeRules.tex [hunk 1]
```diff
@@ -1,13 +1,373 @@
...
-%introduction
+HCL employs types.
+This is done for three reasons:
+\begin{itemize}
+	\item To avoid certain runtime errors, which would not be caught without types.
+	\item Easier allocation of memory, as each type has a pre-defined size.
+	\item As described in the problem analysis, types make it easier for novices to understand constructions such as function definitions.
+\end{itemize}
+The code snippet shown in listing \ref{lis:TypeEx} is a syntactically correct assignment in HCL.
+If type rules were not enforced, this program would fail (or run with unexpected results), as the second argument for the "$+$" function would evaluate to a value of type bool, and no definition for this function allows bool-type arguments.
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: we don't use past tense

File: report/3.Theory/TypeRules.tex [hunk 2]
```diff
@@ -1,13 +1,373 @@
...
+\end{itemize}
+
+Essentially, the type rules specify when certain expressions have to evaluate to specific types.
+The rules have been split up in two sections:
+\begin{itemize}
+	\item Rules for expressions
+	\item Rules for statements
+\end{itemize}
+The former specifies which type an expression evaluates to, depending on the input.
+The latter specifies what types the components of a statement must be to be type valid.
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: "must be, to be type valid" (comma)

File: report/3.Theory/TypeRules.tex [hunk 3]
```diff
@@ -1,13 +1,373 @@
...
+\end{center}
+This is the type rule for addition of two numbers.
+It means that given the current environment, the resulting value will be of type $num$ if the two input expressions are both of type $num$.
+
+As the type rules do not change the state of a program, and therefore also do not change the variable- and function environments, the environment can be simplified to one symbol, $E$.
+The colon is used to show that an expression is of a certain type.
 
 \subsection{Type Rules for Expressions}
-%quick introduction.
+As mentioned earlier, an expression evaluates to a type, if certain premises are upheld (where such apply).
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: omvendt ordstilling. 
"if certain premises are upheld, an expression evaluates to a type."
hele "where such apply" virker ligegyldigt her. not sure

File: report/3.Theory/TypeRules.tex [hunk 4]
```diff
@@ -1,13 +1,373 @@
...
+A $txt$ literal is always of type $txt$.
+
+\begin{center}
+	\begin{math}
+	E \vdash t : txt
+	\end{math}
+\end{center}
+
+\textbf{List literal type}\\
+Every entry in any given list must be of the same type, as long as that type isn't $none$.
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: "but cannot be of type none".. the andet implyer at du godt må have forskellige typer hvis den ene type er none

File: report/3.Theory/TypeRules.tex [hunk 5]
```diff
@@ -1,13 +1,373 @@
...
+		.
+		\ , \ e_k] : list[T]}
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $T \notin \{none\}$
+	\texttt{and} $k > 0$
+\end{center}
+
+\textbf{Tuple literal type}\\
+The tuple type allows each entry to have any type that isn't $none$ or $generic$.
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: De kan godt være generic, så frem den generic er defineret i scopet :)

File: report/3.Theory/TypeRules.tex [hunk 6]
```diff
@@ -1,13 +1,373 @@
...
+
+
 
 \subsection{Type Rules for Statements}
 %quick introduction.
+The type rules for statements are, as mentioned earlier, different from the rules for expressions.
+Statements are not evaluated to a type.
+Rather, they only determine if a program is type valid.
+If a statement is type valid, it is evaluated to \texttt{ok}.
+Otherwise, the compiler with throw an error.
```
Comment (jones1994) [inline report/3.Theory/TypeRules.tex]: change "with thorw" -> "throws"

File: report/3.Theory/TypeRules.tex [hunk 7]
```diff
@@ -1,13 +1,373 @@
...
+
+Just like the rules for expressions, the type rules for statements have been simplified to not include any function calls.
```
Comment (C0DK) [inline report/3.Theory/TypeRules.tex]: hvorfor er de simplified though?

File: report/3.Theory/TypeRules.tex (hunk 3 above)
Comment (Lambcreek) [reply report/3.Theory/TypeRules.tex]: Jeg er virkelig ikke glad for at starte en subsection med "if".

File: report/3.Theory/TypeRules.tex (hunk 7 above)
Comment (Lambcreek) [reply report/3.Theory/TypeRules.tex]: Måske er "simplified" det forkerte ord. Det betyder bare, at vi ikke har specifikke typeregler for fx kald af then, while, etc., fordi deres regler allerede er dækket af reglen for funktionskald

File: report/3.Theory/TypeRules.tex (hunk 3 above)
Comment (Lambcreek) [reply report/3.Theory/TypeRules.tex]: Jeg har skrevet "if and only if" i stedet for "if"

**Pull request merged**